// @astrojs/image@0.4.0/dist/utils/paths.js
import { shorthash } from "./shorthash.js";
function isRemoteImage(src) {
  return /^http(s?):\/\//.test(src);
}
function removeQueryString(src) {
  const index = src.lastIndexOf("?");
  return index > 0 ? src.substring(0, index) : src;
}
function extname(src, format) {
  const index = src.lastIndexOf(".");
  if (index <= 0) {
    return void 0;
  }
  return src.substring(index);
}
function removeExtname(src) {
  const index = src.lastIndexOf(".");
  if (index <= 0) {
    return src;
  }
  return src.substring(0, index);
}
function basename(src) {
  return src.replace(/^.*[\\\/]/, "");
}
function propsToFilename(transform) {
  var _a;
  let filename = removeQueryString(transform.src);
  filename = basename(filename);
  filename = removeExtname(filename);
  const ext = transform.format || ((_a = extname(transform.src)) == null ? void 0 : _a.substring(1));
  return `/${filename}_${shorthash(JSON.stringify(transform))}.${ext}`;
}
function appendForwardSlash(path) {
  return path.endsWith("/") ? path : path + "/";
}
function prependForwardSlash(path) {
  return path[0] === "/" ? path : "/" + path;
}
function removeTrailingForwardSlash(path) {
  return path.endsWith("/") ? path.slice(0, path.length - 1) : path;
}
function removeLeadingForwardSlash(path) {
  return path.startsWith("/") ? path.substring(1) : path;
}
function trimSlashes(path) {
  return path.replace(/^\/|\/$/g, "");
}
function isString(path) {
  return typeof path === "string" || path instanceof String;
}
function joinPaths(...paths) {
  return paths.filter(isString).map(trimSlashes).join("/");
}
export {
  appendForwardSlash,
  isRemoteImage,
  joinPaths,
  prependForwardSlash,
  propsToFilename,
  removeLeadingForwardSlash,
  removeTrailingForwardSlash,
  trimSlashes
};
